### PR TITLE
Backport to Ginkgo

### PR DESCRIPTION
**Pull request by ciuin**
This branch contains changes necessary to run edx-learndot under Open edX Ginkgo.

**Testing instructions**:

1. Set up a Gingko devstack. 
2. Add 'edxlearndot' to settings.INSTALLED_APPS.
3. Add a valid Learndot API key to settings.LEARNDOT_API_KEY.
4. Add a valid Learndot API sandbox URL to settings.LEARNDOT_API_BASE_URL.
5. Create a course you can pass. A new course with one gradable unit that's 100% of the grade would work.
6. Configure a CourseMapping in the Django admin, using a valid Learndot component ID and the course key from the course to be passed.
7. Enroll a learner in the course, complete the gradable unit as that learner, and check that the Learndot enrolment status was set to "PASSED". This will be in the devstack LMS logs, and can also be verified via the Learndot API, using cURL.

**Reviewers**
- [ ] @kaizoku

### CODE REVIEW

File: edxlearndot/learndot.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+class LearndotAPIClient(LearndotAPIClientBase):
+    """
+    Client for the live Learndot API.
+    """
+
+    def get_api_key(self):
+        """
+        Returns the API key for the Learndot v2 API.
+        """
+        return settings.LEARNDOT_API_KEY
```
Comment (kaizoku) [inline edxlearndot/learndot.py]: When I only had the key in `lms.auth.json`, this was throwing an exception.
Not sure if this is just from how my devstack is setup, but I had to actually set it in `lms/env/common.py`.
Same error came with from the `settings.LEARNDOT_API_BASE_URL` below.

File: edxlearndot/signals.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+from openedx.core.djangoapps.signals.signals import COURSE_GRADE_NOW_PASSED
+
+from edxlearndot.learndot import EnrolmentStatus, LearndotAPIClient, LearndotAPIException
+from edxlearndot.models import CourseMapping
+
+
+log = logging.getLogger(__name__)
+
+log.info('edxlearndot.signals imported')
```
Comment (kaizoku) [inline edxlearndot/signals.py]: nit: This shows up a few times on devstack startup and might be a bit verbose.

**Comment by kaizoku** [comment]
The signal fires when the course completes and marks my status as complete in learndot.
I ran into a couple issues along the way and left comments inline about them.

I'm not too sure how to run the tests for this.
I tried `./manage.py test`, but I had to change `DJANGO_SETTINGS_MODULE` to `testing.testsettings`, but they still failed.

File: edxlearndot/signals.py (hunk 2 above)
Comment (ciuin) [reply edxlearndot/signals.py]: Removed.

File: edxlearndot/learndot.py (hunk 1 above)
Comment (ciuin) [reply edxlearndot/learndot.py]: Yeah, the JSON values would need to be pulled from the *TOKENS variables, which I hadn't done. I've changed the README to just say to set them, as I didn't think it was appropriate to add app-specific handling of that under lms/env.

**Comment by ciuin** [comment]
I've added test invocation and clarified settings in the README, and removed the superfluous logging.

**Comment by kaizoku** [comment]
nfortunately when I was testing though, I accidentally created 2 enrollments by running my curl command twice:
```bash
alias lcurl='curl -H "TrainingRocket-Authorization: <redacted>" -H "Content-Type: application/json" -H "Accept: application/json"'
lcurl -s -d '{"contactId": "33822", "id": "122", "componentId": "5", "status": "TENTATIVE"}' https://cloudera-sandbox.trainingrocket.com/api/rest/v2/manage/enrolment | jq -C . | less
lcurl -s -d '{"contactId": "33822", "id": "122", "componentId": "5", "status": "TENTATIVE"}' https://cloudera-sandbox.trainingrocket.com/api/rest/v2/manage/enrolment | jq -C . | less
```

When I completed a course when both of these enrollments existed, it didn't throw any error at first. I added a traceback wrapper around everything and got this traceback:
```
Traceback (most recent call last):
  File "/edx/src/edx-learndot/edxlearndot/signals.py", line 49, in listen_for_passing_grade
    enrolment_id = learndot_client.get_enrolment_id(contact_id, cm.learndot_component_id)
  File "/edx/src/edx-learndot/edxlearndot/learndot.py", line 243, in get_enrolment_id
    enrolment_id = sort_enrolments_by_expiry(enrolments)[-1]["id"]
  File "/edx/src/edx-learndot/edxlearndot/learndot.py", line 52, in sort_enrolments_by_expiry
    return sorted(enrolment_list, key=lambda e: dateutil.parser.parse(e.get("expiryDate")))
  File "/edx/src/edx-learndot/edxlearndot/learndot.py", line 52, in <lambda>
    return sorted(enrolment_list, key=lambda e: dateutil.parser.parse(e.get("expiryDate")))
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/dateutil/parser.py", line 720, in parse
    return DEFAULTPARSER.parse(timestr, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/dateutil/parser.py", line 308, in parse
    res = self._parse(timestr, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/dateutil/parser.py", line 356, in _parse
    l = _timelex.split(timestr)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/dateutil/parser.py", line 150, in split
    return list(cls(s))
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/dateutil/parser.py", line 147, in next
    return self.__next__()  # Python 2.x support
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/dateutil/parser.py", line 141, in __next__
    token = self.get_token()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/dateutil/parser.py", line 72, in get_token
    nextchar = self.instream.read(1)
AttributeError: 'NoneType' object has no attribute 'read'
```

Perhaps we should add more error checking in `sort_enrolments_by_expiry()` ?
Seems like this should be caught [here in `get_enrolment_id()`](https://github.com/open-craft/edx-learndot/blob/47cdf701a685d7fefba9992be472cf72c7ed4851/edxlearndot/learndot.py#L253), but we're just getting a different exception than that handler is expecting.

**Comment by ciuin** [comment]
@kaizoku Thanks again. The root cause was that I wasn't properly handling enrolments without expiry dates. I've overhauled the sorting, added more tests for it, and added better error logging throughout the module.

**Comment by kaizoku** [comment]
:+1:

- [✓] I tested this: (installed in a devstack, created a learndot enrollment, completed a course and checked that learndot enrollment was marked as completed, also ran the python tests)
- [✓] I read through the code
- [n/a] I checked for accessibility issues
- [✓] Includes documentation

Thanks for working through these last few logging issues here.
There's a lot of possible edge cases here with this one.
I think it will really help the ease of debugging down the line if anything winds up going wrong though.

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  pip install -e /edx/src/edx-learndot
+
+*************
+Configuration
+*************
+
+Django configuration
+--------------------
+
+Add ``edxlearndot`` to ``settings.INSTALLED_APPS``.
```
Comment (pomegranited) [inline README.rst]: Worth mentioning that this can also be done via ansible configuration: 
```yaml
EDXAPP_ENV_EXTRA:
  ADDL_INSTALLED_APPS:
    - edxlearndot
```

File: README.rst [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+This enables your edX installation to interact with the `Learndot v2
+API`_.
+
+There are two Django settings you'll need to add:
+
+* Put your Learndot Enterprise base URL (e.g. ``https://your-sandbox.trainingrocket.com/``) in
+  ``settings.LEARNDOT_API_BASE_URL``.
+
+* Put your Learndot API key in ``settings.LEARNDOT_API_KEY``.
```
Comment (pomegranited) [inline README.rst]: If you include the `plugin_settings` as [suggested below](https://github.com/open-craft/edx-learndot/pull/2/files#diff-9cd30b8f8b6ed0a35c55764479e355dfR23), then it's worth mentioning that this can be done by adding `LEARNDOT_API_BASE_URL` to the `lms.env.json`, and `LEARNDOT_API_KEY` to lms.auth.json.

File: edxlearndot/apps.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+"""
+
+from __future__ import absolute_import, unicode_literals
+
+import logging
+
+from django.apps import AppConfig
+
+
+log = logging.getLogger(__name__)
```
Comment (pomegranited) [inline edxlearndot/apps.py]: nit: No logging done here anymore -- please remove this and the `import logging` above.

File: edxlearndot/apps.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+class LearndotIntegrationConfig(AppConfig):
+    """
+    Configures edxlearndot as a Django app
+    """
+    name = 'edxlearndot'
+    verbose_name = "edX Learndot Integration"
+
+    def ready(self):
+        import edxlearndot.signals  # pylint: disable=unused-variable
```
Comment (pomegranited) [inline edxlearndot/apps.py]: Since we don't want to have to alter the LMS env code for our client, could you please include the [`edxlearndot.settings`](https://github.com/open-craft/edx-learndot/pull/1/files#diff-27d9a09c4f4808c570368ce3984f6a3a) file in this branch, and invoke it here to update the settings from the env?  E.g. 
```python
        from edxlearndot.settings import plugin_settings
        plugin_settings(settings)
```

Will also need to add `from django.conf import settings` above.

File: get_edx_platform.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+WORKING_COPY="${ROOT}/.tox/edx-platform"
+
+function check_repo_remote() {
+    (cd "$WORKING_COPY" && git remote -v | grep "$REPO")
+}
+
+function clone() {
+    echo "Cloning edx-platform from $REPO, branch $BRANCH"
+    rm -rf "$WORKING_COPY"
+    git clone -b "$BRANCH" "$REPO" "$WORKING_COPY"
```
Comment (pomegranited) [inline get_edx_platform.sh]: To make this clone run faster, could you add an optional reference copy?  e.g.

```bash
REFERENCE_COPY="${OPENEDX_COPY:-/edx/app/edxapp/edx-platform}"

git clone -b "$BRANCH" "$REPO" "$WORKING_COPY" --reference-if-able "$REFERENCE_COPY"
```

Comment (ciuin) [reply get_edx_platform.sh]: Added.

File: edxlearndot/apps.py (hunk 6 above)
Comment (ciuin) [reply edxlearndot/apps.py]: Done.

File: edxlearndot/apps.py (hunk 5 above)
Comment (ciuin) [reply edxlearndot/apps.py]: Done.

**Comment by ciuin** [comment]
I'm closing the PR, as we're not merging this to master, but keeping the ginkgo branch to support that release.
